yoyodyne/models/modules/lstm.py

```diff
@@ -223,3 +223,167 @@ def forward(
     @property
     def name(self) -> str:
         return "attentive LSTM"
+
+
+class HMMLSTMDecoder(LSTMDecoder):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.output_proj = nn.Sequential(
+            nn.Linear(self.output_size, self.output_size), nn.Tanh()
+        )
+        self.scale_encoded = nn.Linear(
+            self.decoder_input_size, self.hidden_size
+        )
+
+    def _alignment_step(self, decoded, encoder_out, encoder_mask):
+        # Multiplies encoding and decoding for alignment representations.
+        # -> B x seq_len.
+        alignment_scores = torch.bmm(
+            self.scale_encoded(encoder_out), decoded.transpose(1, 2)
+        ).squeeze(-1)
+        # Gets probability of alignments.
+        alignment_probs = nn.functional.softmax(alignment_scores, dim=-1)
+        # Mask padding.
+        alignment_probs = alignment_probs * (~encoder_mask) + 1e-7
```

Comment:
Why can't we do normal masking here? That is, I understood the smoothing in hmm.py to be for enforcing monotonicity in a differentiable way, but I don't think that's what were doing here?

Reply:
It doesn't work here. I don't know why, but this is what they had to do in the original code. Everytime I replace with normal masking it breaks. I'm just sticking to this for the first commit and cleaning it up once I have a moment to stare at tensor values.

Reply:
Hmmm I would be interested to know why it doesn't work (though I would need to dig back through the code to have any opinion anyway, so dont let me hold up the PR here.)

Reply:
No it's weird. It's next up on the debug because it's annoying that it doesn't work.